### ISSUE TITLE
Errors when firing a signal just after awaiting on that signal

### ISSUE THREAD

**Issue by BimDav** [godotengine/godot#89439]
### Tested versions

4.2.stable

### System information

MacOS, M2

### Issue description

Errors are logged when firing a signal just after awaiting on that signal. It seems that the await object is not properly cleaned up and is still partly waiting on the signal.

```
extends Node2D

signal my_signal()

func _ready() -> void:
	await my_signal
	my_signal.emit()

func _process(_delta: float) -> void:
	my_signal.emit()
	set_process(false)
```

gives 

```
E 0:00:00:0589   Node2D.gd:8 @ _ready(): Resumed function '_ready()' after await, but script is gone. At script: res://Node2D.gd:7
  <Erreur C++>   Method/function failed. Returning: Variant()
  <Source C++>   modules/gdscript/gdscript_function.cpp:197 @ resume()
  <Pile des appels>Node2D.gd:8 @ _ready()
                 Node2D.gd:13 @ _process()

E 0:00:00:0589   Node2D.gd:13 @ _process(): Attempt to disconnect a nonexistent connection from 'Node2D:<Node2D#26793215177>'. Signal: 'my_signal', callable: 'GDScriptFunctionState::_signal_callback'.
  <Erreur C++>   Condition "!s->slot_map.has(*p_callable.get_base_comparator())" is true. Returning: false
  <Source C++>   core/object/object.cpp:1420 @ _disconnect()
  <Pile des appels>Node2D.gd:13 @ _process()
```

### Steps to reproduce

Run the project below or the code sample above.

### Minimal reproduction project (MRP)

[await_signal_twice_fired.zip](https://github.com/godotengine/godot/files/14585959/await_signal_twice_fired.zip)

**Comment by ivoiv** [comment]
Aren't you doing something "illegal"?

You're blocking your `_ready` from executing before `_process` has emitted the signal.
set_process docs say and probably for good reason
```
Any calls to this before _ready will be ignored.
```

Whatever called the `_ready` is probably stuck waiting on the `await` as well so maybe it can't clean up or assign things properly?

**Comment by BimDav** [comment]
It's just a minimal code example to reproduce the bug. The bug is not linked to this happening in `_ready` or not. The bug is also not linked to the `set_process(false)`, which I added to remove confusion due to the signal being emitted each frame. This way, it is only emitted once.

**Comment by kleonc** [comment]
The direct cause is that `ONE_SHOT` connections to a signal are disconnected after the connected method is called, not before. Meaning e.g. this results in an infinite loop:
```gdscript
extends Node2D

signal my_signal

func _ready():
	my_signal.connect(foo, CONNECT_ONE_SHOT)
	my_signal.emit()

func foo():
	my_signal.emit()
```

Related issue: #28857.

How does it affect `await`? When awaiting on a signal `GDScriptFunctionState::_signal_callback` which calls `GDScriptFunctionState::resume` is `ONE_SHOT`-connected to such awaited signal:

https://github.com/godotengine/godot/blob/61282068f4d59cb48f35ad95391728c58d9008ab/modules/gdscript/gdscript_vm.cpp#L2306
https://github.com/godotengine/godot/blob/61282068f4d59cb48f35ad95391728c58d9008ab/modules/gdscript/gdscript_function.cpp#L137-L167

Meaning if the awaited signal is emitted again during such `resume()` then it results in calling `resume()` again because the `ONE_SHOT` connection is not yet disconnected.

---
```gdscript
extends Node2D

signal my_signal

func _ready() -> void:
	foo()
	my_signal.emit() # A

func foo():
	await my_signal
	my_signal.emit() # B
```
Execution goes something like:
```
`_ready()`
	`foo()`
		`await my_signal`
			`ONE_SHOT` connect resuming `foo` on `my_signal`
	`my_signal.emit() # A`
		Resume `foo`.
			`my_signal.emit() # B`
				Resume `foo`.
					Error (already removed from `scripts_list` within the first resume).
				Disconnect resuming `foo` on `my_signal` signal (it's ONE_SHOT and `my_signal.emit() # B` handled all connections).
			foo "completed" signal
		Error trying to disconnect resuming `foo` on `my_signal` signal (`my_signal.emit() # A` handled all connections).
```
The in `GDScriptFunctionState::resume` error part:
https://github.com/godotengine/godot/blob/61282068f4d59cb48f35ad95391728c58d9008ab/modules/gdscript/gdscript_function.cpp#L190-L212

---

How disconnecting `ONE_SHOT` connections is done:
https://github.com/godotengine/godot/blob/61282068f4d59cb48f35ad95391728c58d9008ab/core/object/object.cpp#L1121-L1175

**Comment by BimDav** [comment]
Thanks for investigating! @AThousandShips it is still a bug, though, right?

**Comment by BimDav** [comment]
That's what I get for giving a minimal reproducing example :). 

Consider this: I want to wait for the state to get back to `Idle` before updating it to `State1`. There is a `set_state()` setter and a `state_changed` signal:
```
func set_state(p_state: State) -> void:
   state = p_state
   state_changed.emit()
```

now in my updating function:

```
func wait_and_update_to_state1() -> void:
   while state != Idle:
      await state_changed
   state = State1
```

This seems reasonable, right? It gives the above errors.

**Comment by AThousandShips** [comment]
> That's what I get for giving a minimal reproducing example

What do you mean?

As explained above awaiting uses connections, so that's an unfortunate consequence of the same limitation, but I don't think it's solvable to handle this as such, try using deferred calls to update the data perhaps, otherwise I'd say this is a limitation just like call deferred

**Comment by BimDav** [comment]
The problem of giving minimal reproducing example is that it can often be answered by "why would even do that anyway?". I should have given more realistic context from the start.

In my opinion, it's not because something is complicated to fix that it's not a bug: I think the above example is a perfectly reasonable usecase for await, and it not working without weird errors being logged constitutes a bug.

**Comment by AThousandShips** [comment]
> In my opinion, it's not because something is complicated to fix that it's not a bug

Agreed, I never said it wasn't a bug because of that, please read what I say instead of putting words in my moth, act professionally and please read the [code of conduct](https://godotengine.org/code-of-conduct)

I'm going to stop engaging with you and I'll wish you a good day

**Comment by kleonc** [comment]
@AThousandShips
To clarify: I've only pointed out the direct cause of the reported issue for the current implementation.

Whether how `ONE_SHOT` connections are being disconnected is intentional or it is a bug itself I don't know, hence why [I've asked about this in the contributors chat](https://chat.godotengine.org/channel/core?msg=kFQYGzuXwBAdojfLN) (I could have mentioned my doubts in my previous comment).

But regardless of the current `ONE_SHOT` disconnecting behavior, what's reported here i.e. `await` producing errors was already triaged as a bug by a GDScript team member (**dalexeev**) so please respect this more / be more thoughful about changing the labels in the future (I'll leave fixing them to you).

The behavior of `await` seems already correct / as expected, what's reported here is some superfluous/unneeded errors.

---

So:
- If the current `ONE_SHOT` disconnecting behavior is a bug then it needs to be fixed. The issue with `await` erroring would then be automatically fixed as a result.
- If the current `ONE_SHOT` disconnecting behavior is not a bug then `await`'s implementation needs to be tweaked to fix the reported issue accordingly.

---

@BimDav Your original MRP is fine/sufficient on its own, no need for a "real world" example to classify some easily reproducible behavior as a bug / not a bug.

**Comment by AThousandShips** [comment]
And I gave my arguments in chat too, and explained here, but I'll restore the label, but I thought this through and presented my arguments, so please take that into account :)

**Comment by kleonc** [comment]
_(All below responses are not about the subject of the issue, hence I'm marking this as offtopic myself.)_

@AThousandShips

> And I gave my arguments in chat too, and explained here

Great, thanks!

>, but I'll restore the label, but I thought this through and presented my arguments, so please take that into account

Sure I take this into account. But I also take into account that others think things through as well (here "labelers"). What I mean is if I see that GDScript/rendering/etc. team member/maintainer marked an issue in the given area as a bug then in case I disagree I'd state my arguments and let them reevaluate the issue. I wouldn't unmark it as a bug on my own because I think it's not a bug etc. (unless that's something really obvious/trivial; not the case in here). Maybe I'm right that it's not a bug and they'd agree with me (and change labels). But maybe I'm not aware of something else and it is indeed a bug and it's me still not taking something into account. What I'm saying is that in such situations there's no need to rush and "take over" by changing the labels. I think we both agree that we don't want labels being changed back and forth whenever some member have different opinion about them.
And if this needs clarification: I'm not saying any of this to attack you or something like that. I think that's something you could improve on, hence my suggestion in my previous comment. That's all. :slightly_smiling_face:

>(and don't misrepresent my actions) :)

I don't think I did, for sure I haven't tried to do so. So I'm not promising anything. :upside_down_face:

> also note that the original labelling of this was made without the knowledge of these details, so it's not irrelevant

Now that's just your assumption that **dalexeev** wasn't aware of these details. And nothing wrong with this! Making some assumptions about others is of course fine.
And while I agree you're likely right that he haven't dived deeply into the details at the time of labelling things, I still stand my point - I think it would be better if you'd restrained yourself from changing the labels in a situation like in here.

I don't expect any response to this from you, I'd say let's not continute this offtopic about meta stuff (or at least not here). That's just my opinion and if you take anything positive from it - great. :slightly_smiling_face:

**Comment by BimDav** [comment]
I have a second, similar but different enough case. Here is a sample project: 
[resume_after_await.zip](https://github.com/godotengine/godot/files/14603203/resume_after_await.zip)

```
extends Node2D

signal signal1
signal signal2


func _ready() -> void:
	signal1.connect(_on_signal1_emitted)
	var coroutine: = Coroutine.new()
	coroutine.add_future(signal1)   # coroutine awaits on signal1
	coroutine.add_future(signal2)   # coroutine awaits on signal2
	await coroutine.join_either()   # signal1 will resume execution here and coroutine will be deleted
	print("after")


func _process(delta: float) -> void:
	signal1.emit()


func _on_signal1_emitted() -> void:
	signal2.emit()  # this gives an error as coroutine is deleted
```

The issue here seems to be that the receivers list stills contains `coroutine`'s `await` even though it was just deleted.
